File: web_api_tests/python/concurrent_test.py
import asyncio
import aiohttp
import time

BASE_URL = "http://localhost:3001"

async def make_request(session, method, path, json=None):
    async with session.request(method, f"{BASE_URL}{path}", json=json) as resp:
        try:
            data = await resp.json()
        except:
            data = None
        return {"status": resp.status, "data": data}

async def run_concurrent_test():
    print("\n========================================")
    print("  PYTHON/FASTAPI - CONCURRENT TEST")
    print("========================================\n")

    start_time = time.time()

    async with aiohttp.ClientSession() as session:
        # Setup data
        print("Setting up test data...")
        user_ids = []
        for i in range(50):
            res = await make_request(session, "POST", "/api/users", {
                "username": f"concurrent{i}",
                "email": f"concurrent{i}@example.com",
                "displayName": f"Concurrent User {i}",
            })
            user_ids.append(res["data"]["id"])

        post_ids = []
        for i in range(100):
            user_id = user_ids[i % len(user_ids)]
            res = await make_request(session, "POST", "/api/posts", {
                "userId": user_id,
                "content": f"Post {i}",
            })
            post_ids.append(res["data"]["id"])

        # Concurrent Test 1: Parallel reads
        print("Concurrent Test 1: Parallel Reads (200 simultaneous GETs)...")
        t1 = time.time()
        tasks1 = []
        for i in range(200):
            user_id = user_ids[i % len(user_ids)]
            tasks1.append(make_request(session, "GET", f"/api/users/{user_id}"))
        await asyncio.gather(*tasks1)
        t1_time = (time.time() - t1) * 1000
        print(f"  ✓ Completed in {t1_time:.0f}ms")

        # Concurrent Test 2: Parallel writes
        print("Concurrent Test 2: Parallel Writes (100 simultaneous POSTs)...")
        t2 = time.time()
        tasks2 = []
        for i in range(100):
            user_id = user_ids[i % len(user_ids)]
            tasks2.append(make_request(session, "POST", "/api/posts", {
                "userId": user_id,
                "content": f"Concurrent post {i}",
            }))
        await asyncio.gather(*tasks2)
        t2_time = (time.time() - t2) * 1000
        print(f"  ✓ Completed in {t2_time:.0f}ms")

        # Concurrent Test 3: Mixed operations
        print("Concurrent Test 3: Mixed Operations (300 mixed requests)...")
        t3 = time.time()
        tasks3 = []
        for i in range(300):
            op = i % 4
            if op == 0:
                # POST comment
                tasks3.append(make_request(session, "POST", "/api/comments", {
                    "postId": post_ids[i % len(post_ids)],
                    "userId": user_ids[i % len(user_ids)],
                    "text": f"Mixed comment {i}",
                }))
            elif op == 1:
                # GET feed
                tasks3.append(make_request(session, "GET", f"/api/users/{user_ids[i % len(user_ids)]}/feed"))
            elif op == 2:
                # POST like
                tasks3.append(make_request(session, "POST", "/api/likes", {
                    "postId": post_ids[i % len(post_ids)],
                    "userId": user_ids[i % len(user_ids)],
                }))
            else:
                # POST follow
                tasks3.append(make_request(session, "POST", "/api/follow", {
                    "followerId": user_ids[i % len(user_ids)],
                    "followingId": user_ids[(i + 1) % len(user_ids)],
                }))
        await asyncio.gather(*tasks3, return_exceptions=True)
        t3_time = (time.time() - t3) * 1000
        print(f"  ✓ Completed in {t3_time:.0f}ms")

    total_time = (time.time() - start_time) * 1000

    print("\n========================================")
    print("         CONCURRENT TEST RESULTS")
    print("========================================")
    print(f"Test 1 - Parallel Reads:     {t1_time:.0f}ms")
    print(f"Test 2 - Parallel Writes:    {t2_time:.0f}ms")
    print(f"Test 3 - Mixed Operations:   {t3_time:.0f}ms")
    print(f"Total Execution Time:        {total_time:.0f} ms")
    print("========================================\n")

if __name__ == "__main__":
    asyncio.run(run_concurrent_test())

File: web_api_tests/csharp/ConcurrentTest.cs
using System;
using System.Net.Http;
using System.Collections.Generic;
using System.Threading.Tasks;
using System.Diagnostics;

class ConcurrentTest
{
    private static readonly string BASE_URL = "http://localhost:3002";
    private static readonly HttpClient client = new();

    static async Task Main()
    {
        Console.WriteLine("\n========================================");
        Console.WriteLine("  C#/.NET - CONCURRENT TEST");
        Console.WriteLine("========================================\n");

        var startTime = Stopwatch.GetTimestamp();

        // Setup data
        Console.WriteLine("Setting up test data...");
        var userIds = new List<int>();
        for (int i = 0; i < 50; i++)
        {
            var res = await MakeRequest("POST", "/api/users", new
            {
                username = $"concurrent{i}",
                email = $"concurrent{i}@example.com",
                displayName = $"Concurrent User {i}",
            });
            if (res != null && res.ContainsKey("id"))
                userIds.Add((int)(long)res["id"]);
        }

        var postIds = new List<int>();
        for (int i = 0; i < 100; i++)
        {
            var res = await MakeRequest("POST", "/api/posts", new
            {
                userId = userIds[i % userIds.Count],
                content = $"Post {i}",
            });
            if (res != null && res.ContainsKey("id"))
                postIds.Add((int)(long)res["id"]);
        }

        // Concurrent Test 1: Parallel reads
        Console.WriteLine("Concurrent Test 1: Parallel Reads (200 simultaneous GETs)...");
        var t1 = Stopwatch.GetTimestamp();
        var tasks1 = new List<Task>();
        for (int i = 0; i < 200; i++)
        {
            tasks1.Add(client.GetAsync($"{BASE_URL}/api/users/{userIds[i % userIds.Count]}"));
        }
        await Task.WhenAll(tasks1);
        var t1Time = Stopwatch.GetElapsedTime(t1).TotalMilliseconds;
        Console.WriteLine($"  ✓ Completed in {t1Time:F0}ms");

        // Concurrent Test 2: Parallel writes
        Console.WriteLine("Concurrent Test 2: Parallel Writes (100 simultaneous POSTs)...");
        var t2 = Stopwatch.GetTimestamp();
        var tasks2 = new List<Task>();
        for (int i = 0; i < 100; i++)
        {
            tasks2.Add(MakeRequest("POST", "/api/posts", new
            {
                userId = userIds[i % userIds.Count],
                content = $"Concurrent post {i}",
            }));
        }
        await Task.WhenAll(tasks2);
        var t2Time = Stopwatch.GetElapsedTime(t2).TotalMilliseconds;
        Console.WriteLine($"  ✓ Completed in {t2Time:F0}ms");

        // Concurrent Test 3: Mixed operations
        Console.WriteLine("Concurrent Test 3: Mixed Operations (300 mixed requests)...");
        var t3 = Stopwatch.GetTimestamp();
        var tasks3 = new List<Task>();
        for (int i = 0; i < 300; i++)
        {
            int op = i % 4;
            if (op == 0)
            {
                tasks3.Add(MakeRequest("POST", "/api/comments", new
                {
                    postId = postIds[i % postIds.Count],
                    userId = userIds[i % userIds.Count],
                    text = $"Mixed comment {i}",
                }));
            }
            else if (op == 1)
            {
                tasks3.Add(client.GetAsync($"{BASE_URL}/api/users/{userIds[i % userIds.Count]}/feed"));
            }
            else if (op == 2)
            {
                tasks3.Add(MakeRequest("POST", "/api/likes", new
                {
                    postId = postIds[i % postIds.Count],
                    userId = userIds[i % userIds.Count],
                }));
            }
            else
            {
                tasks3.Add(MakeRequest("POST", "/api/follow", new
                {
                    followerId = userIds[i % userIds.Count],
                    followingId = userIds[(i + 1) % userIds.Count],
                }));
            }
        }
        await Task.WhenAll(tasks3);
        var t3Time = Stopwatch.GetElapsedTime(t3).TotalMilliseconds;
        Console.WriteLine($"  ✓ Completed in {t3Time:F0}ms");

        var totalTime = Stopwatch.GetElapsedTime(startTime).TotalMilliseconds;

        Console.WriteLine("\n========================================");
        Console.WriteLine("         CONCURRENT TEST RESULTS");
        Console.WriteLine("========================================");
        Console.WriteLine($"Test 1 - Parallel Reads:     {t1Time:F0}ms");
        Console.WriteLine($"Test 2 - Parallel Writes:    {t2Time:F0}ms");
        Console.WriteLine($"Test 3 - Mixed Operations:   {t3Time:F0}ms");
        Console.WriteLine($"Total Execution Time:        {totalTime:F0} ms");
        Console.WriteLine("========================================\n");
    }

    static async Task<Dictionary<string, object>> MakeRequest(string method, string path, object body)
    {
        try
        {
            var request = new HttpRequestMessage(new HttpMethod(method), $"{BASE_URL}{path}");
            if (body != null)
            {
                var json = System.Text.Json.JsonSerializer.Serialize(body);
                request.Content = new StringContent(json, System.Text.Encoding.UTF8, "application/json");
            }
            var response = await client.SendAsync(request);
            var content = await response.Content.ReadAsStringAsync();
            return System.Text.Json.JsonSerializer.Deserialize<Dictionary<string, object>>(content);
        }
        catch { return null; }
    }
}
